### ISSUE TITLE
GetAllGlobs are quite slow

### ISSUE THREAD

**Issue by lifengl** [Microsoft/msbuild#2421]
I took a perf trace in the project loading scenario (load Roslyn project into VS), and GetAllGlobs uses considerable CPU resources.  (In my trace it took about 1.4 seconds)

```
Name | Inc % | Inc
-- | -- | --
Microsoft.Build!Project.GetAllGlobs | 2.5 | 1,388
+ Microsoft.Build!Project.BuildGlobResultFromIncludeItem | 1.7 | 934
\|+ Microsoft.Build!Microsoft.Build.Globbing.CompositeGlob..ctor(class System.Collections.Generic.IEnumerable`1) | 1.5 | 837
\|\|+ System.Collections.Immutable!ImmutableArray.ToImmutableArray | 1.5 | 837
\|\| + System.Collections.Immutable!ImmutableArray.CreateRange | 1.5 | 837
\|\|  + system.core.ni!? | 1.5 | 837
\|\|   + Microsoft.Build!Microsoft.Build.Evaluation.ItemSpec`2+<>c[System.__Canon,System.__Canon].<ToMSBuildGlob>b__19_0(class Microsoft.Build.Evaluation.ItemFragment) | 1.5 | 832
\|\|   \|+ Microsoft.Build!ItemFragment.get_MsBuildGlob | 1.5 | 832
\|\|   \| + mscorlib.ni!System.Lazy`1[System.__Canon].get_Value() | 1.5 | 832
\|\|   \|  + mscorlib.ni!System.Lazy`1[System.__Canon].LazyInitValue() | 1.5 | 831
\|\|   \|  \|+ mscorlib.ni!System.Lazy`1[System.__Canon].CreateValue() | 1.5 | 829
\|\|   \|  \|\|+ Microsoft.Build!ItemFragment.CreateMsBuildGlob | 1.5 | 829
\|\|   \|  \|\| + Microsoft.Build!MSBuildGlob.Parse | 1.5 | 827
\|\|   \|  \|\| \|+ Microsoft.Build!FileMatcher.GetFileSpecInfo | 1.5 | 798
\|\|   \|  \|\| \|\|+ Microsoft.Build!FileMatcher.RegularExpressionFromFileSpec | 1.4 | 788
\|\|   \|  \|\| \|\|\|+ mscorlib.ni!StringBuilder.Replace | 1.3 | 705
\|\|   \|  \|\| \|\|\|\|+ mscorlib.ni!StringBuilder.StartsWith | 0.8 | 455
```

One part of the slowness is coming from RegularExpressionFromFileSpec does lots of string replacements.  May want to investigate how to reduce those overhead, because globbings are now used in every .Net Core projects.

**Comment by davkean** [comment]
This looks related to https://github.com/Microsoft/msbuild/issues/2380, which was dup'd against: https://github.com/Microsoft/msbuild/issues/2392.

**Comment by davkean** [comment]
Dup'ing against #2380, if we reopen it - it will be addressed by it, or by #2392.
